sdk/python/lib/pulumi/x/automation/local_workspace.py

```diff
@@ -0,0 +1,399 @@
+# Copyright 2016-2020, Pulumi Corporation.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import tempfile
+import json
+import yaml
+from concurrent import futures
+from dataclasses import dataclass, asdict
+from typing import Optional, List, Mapping, Callable
+
+from .config import ConfigMap, ConfigValue
+from .project_settings import ProjectSettings
+from .stack_settings import StackSettings
+from .workspace import Workspace, PluginInfo, StackSummary, WhoAmIResult, PulumiFn
+from .stack import Stack
+from .cmd import _run_pulumi_cmd, CommandResult
+
+setting_extensions = [".yaml", ".yml", ".json"]
+
+StackInitializer = Callable[[str, Workspace], Stack]
+
+
+@dataclass
+class LocalWorkspaceOptions:
+    work_dir: Optional[str] = None
+    pulumi_home: Optional[str] = None
+    program: Optional[PulumiFn] = None
+    env_vars: Optional[Mapping[str, str]] = None
+    secrets_provider: Optional[str] = None
+    project_settings: Optional[ProjectSettings] = None
+    stack_settings: Optional[Mapping[str, StackSettings]] = None
+
+
+class LocalWorkspace(Workspace):
+    """
+    LocalWorkspace is a default implementation of the Workspace interface.
+    A Workspace is the execution context containing a single Pulumi project, a program,
+    and multiple stacks. Workspaces are used to manage the execution environment,
+    providing various utilities such as plugin installation, environment configuration
+    ($PULUMI_HOME), and creation, deletion, and listing of Stacks.
+    LocalWorkspace relies on Pulumi.yaml and Pulumi.<stack>.yaml as the intermediate format
+    for Project and Stack settings. Modifying ProjectSettings will
+    alter the Workspace Pulumi.yaml file, and setting config on a Stack will modify the Pulumi.<stack>.yaml file.
+    This is identical to the behavior of Pulumi CLI driven workspaces.
+    """
+
+    @classmethod
+    def new_stack(cls,
```

Comment:
I changed it to `create_new_stack` and `select_existing_stack` and kept `create_or_select_stack`.

Reply:
After getting some use with the API, I decided to extract the class methods on `LocalWorkspace` into standalone functions `create_stack`, `select_stack` and `create_or_select_stack`. The slightly differing names for the methods on LocalWorkspace felt clunky in the IDE with autocomplete.

I noticed that the [Go API](https://github.com/pulumi/pulumi/blob/master/sdk/go/x/auto/local_workspace.go#L712-L902) uses factory functions itself, so the discrepancy between language implementations to align with language culture is already present.

We also have the option of overwriting the class method with an instance-specific version as suggested in [this post](https://stackoverflow.com/a/28766809) but I wasn't really loving that.

I think the standalone functions are pythonic, easy to use, are aligned with the method names in `LocalWorkspace` and `Stack` and `auto.create_stack()` reads better than `auto.LocalWorkspace.create[_new]_stack()`.

Reply:
Notice though, the Go SDK uses `upsert` instead of `create_or_select`. Do we want to do that in python? Or is `create_or_select_stack` the way to go? Another possible option would be to eliminate the 3rd function and have a bool argument to `create_stack`, something like `error_if_exists` that is default `True` and setting to `False` is effectively `create_or_select`.